app/vlselect/traces/query/query.go

```diff
@@ -0,0 +1,404 @@
+func getTraceIDList(ctx context.Context, cp *CommonParams, param *TraceQueryParam) ([]string, time.Time, error) {
+	currentTime := time.Now()
+	// query: * AND <filter> | last 1 by (_time) partition by (trace_id) | fields _time, trace_id | sort by (_time) desc
+	qStr := "*"
+	if param.ServiceName != "" {
+		qStr += fmt.Sprintf("AND _stream:{"+pb.ResourceAttrServiceName+"=\"%s\"} ", param.ServiceName)
+	}
+	if param.SpanName != "" {
+		qStr += fmt.Sprintf("AND _stream:{"+pb.NameField+"=\"%s\"} ", param.SpanName)
+	}
+	if len(param.Attributes) > 0 {
+		for k, v := range param.Attributes {
+			qStr += fmt.Sprintf(`AND "%s":="%s" `, k, v)
+		}
+	}
+	if param.DurationMin > 0 {
+		qStr += fmt.Sprintf("AND "+pb.DurationField+":>%d ", param.DurationMin.Nanoseconds())
+	}
+	if param.DurationMax > 0 {
+		qStr += fmt.Sprintf("AND duration:<%d ", param.DurationMax.Nanoseconds())
+	}
+	qStr += " | last 1 by (_time) partition by (" + pb.TraceIDField + ") | fields _time, " + pb.TraceIDField + " | sort by (_time) desc"
+
+	q, err := logstorage.ParseQueryAtTimestamp(qStr, currentTime.UnixNano())
+	if err != nil {
+		return nil, time.Time{}, fmt.Errorf("cannot parse query [%s]: %s", qStr, err)
+	}
+	q.AddPipeLimit(uint64(param.Limit))
+
+	traceIDs, maxStartTime, err := findTraceIDsSplitTimeRange(ctx, q, cp, param.StartTimeMin, param.StartTimeMax, param.Limit)
+	if err != nil {
+		return nil, maxStartTime, err
+	}
+
+	return traceIDs, maxStartTime, nil
+}
+
+// findTraceIDsSplitTimeRange try to search from the nearest time range of the end time.
+// if the result already met requirement of `limit`, return.
+// otherwise, amplify the time range to 5x and search again, until the start time exceed the input.
+func findTraceIDsSplitTimeRange(ctx context.Context, q *logstorage.Query, cp *CommonParams, startTime, endTime time.Time, limit int) ([]string, time.Time, error) {
+	currentTime := time.Now()
+
+	step := time.Minute
+	currentStartTime := endTime.Add(-step)
+
+	traceIDList := make([]string, 0, limit)
+	maxStartTimeStr := endTime.Format(time.RFC3339)
+
+	writeBlock := func(_ uint, db *logstorage.DataBlock) {
+		columns := db.Columns
+		for i := range columns {
+			if columns[i].Name == "trace_id" {
+				traceIDList = append(traceIDList, columns[i].Values...)
+			} else if columns[i].Name == "_time" {
+				for _, v := range columns[i].Values {
+					if v < maxStartTimeStr {
+						maxStartTimeStr = v
+					}
+				}
+			}
+		}
+	}
+
+	for currentStartTime.After(startTime) {
+		qClone := q.CloneWithTimeFilter(currentTime.UnixNano(), currentStartTime.UnixNano(), endTime.UnixNano())
+		if err := vlstorage.RunQuery(ctx, cp.TenantIDs, qClone, writeBlock); err != nil {
+			return nil, time.Time{}, err
+		}
+
+		// found enough trace_id, return directly
+		if len(traceIDList) == limit {
+			maxStartTime, err := time.Parse(time.RFC3339, maxStartTimeStr)
+			if err != nil {
+				return nil, maxStartTime, err
```

Comment:
I dont think we should return any value in case of an error, as far as I can see returned maxStartTime is not being used anywhere if err != nil

Reply:
Thanks. Updated.